npm/private/npm_translate_lock.bzl

```diff
@@ -506,21 +497,72 @@ Check the public_hoist_packages attribute for duplicates.
                 fail(msg)
 
 def _impl(rctx):
-    lockfile = _process_lockfile(rctx)
+    lockfile = None
+    root_package = None
+    link_workspace = None
+    lockfile_description = None
+
+    if rctx.attr.pnpm_lock != None:
+        if rctx.attr.npm_package_json != None or rctx.attr.npm_package_lock != None:
+            fail("npm_translate_lock requires either pnpm_lock or npm_package_json/npm_package_lock attributes, not both.")
+
+        # don't allow a pnpm lock file that isn't in the root directory of a bazel package
+        if paths.dirname(rctx.attr.pnpm_lock.name):
+            fail("pnpm-lock.yaml file must be at the root of a bazel package")
 
-    # root package is the directory of the pnpm_lock file
-    root_package = rctx.attr.pnpm_lock.package
+        lockfile = _process_lockfile(rctx, rctx.attr.pnpm_lock)
 
-    generated_by_lines = _generated_by_lines(rctx.attr.pnpm_lock.workspace_name, rctx.attr.pnpm_lock)
+        # root package is the directory of the pnpm_lock file
+        root_package = rctx.attr.pnpm_lock.package
+        link_workspace = rctx.attr.pnpm_lock.workspace_name
+
+        lockfile_description = "@{}{}".format(
+            str(rctx.attr.pnpm_lock.workspace_name),
+            str(rctx.attr.pnpm_lock),
+        )
+    else:
+        if rctx.attr.npm_package_json == None or rctx.attr.npm_package_lock == None:
+            fail("npm_translate_lock requires either pnpm_lock or npm_package_json/npm_package_lock attributes.")
+
+        rctx.file(
+            "package.json",
+            content = rctx.read(rctx.attr.npm_package_json),
+            executable = False,
+        )
+        rctx.file(
+            "package-lock.json",
+            content = rctx.read(rctx.attr.npm_package_lock),
+            executable = False,
+        )
+        rctx.execute([
+            rctx.path(Label("@nodejs_host//:bin/node")),
+            rctx.path(Label("@pnpm//:package/bin/pnpm.cjs")),
```

Comment:
I was somewhat surprised that this seemed to work without adding `npm_import(name = "pnpm", package = "pnpm", ...)` anywhere. Is there already a `@pnpm` repository available?

Reply:
yes, it was added here: https://github.com/aspect-build/rules_js/pull/157/files#diff-5493ff8e9397811510e780de47c57abb70137f1afe85d1519130dc3679d60ce5R78

But since that's not a dependency we expose to users, this code will fail under some other workspace, such as in the e2e/ folders.

Reply:
I changed `npm_translate_lock` to a wrapper macro so that it can `maybe()` fetch the `@pnpm` repository in case one of the alternate lockfiles is provided.